regression_em/model.py

```diff
@@ -1,18 +1,287 @@
+"""Package to perform regression EM algorithm self.
+For the detail of this algortithm,
+["Position Bias Estimation for Unbiased Learning to Rank in Personal Search" Xuanhui Wang et al.]
+(https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/46485.pdf)
+"""
+from random import random
+# from typing import Sequence, Tuple, Union
+from typing import Tuple, Union
+
+import numpy as np
 import scipy.sparse as sp
-from sklearn.linear_model import LinearRegression
+from sklearn.linear_model import LinearRegression, Ridge
+from sklearn.base import BaseEstimator
+
+
+# Define type.
+Matrix = Union[np.ndarray, sp.csr_matrix, sp.csc_matrix]
+RegParam = Tuple[Matrix, float]
+
+
+class RegressionEM(BaseEstimator):
+    """Regression EM can estimate latent factors based on below model.
+
+       Outcome = Left latent * Right latent
+
+       For the detail of Regression EM algortithm,
+       ["Position Bias Estimation for Unbiased Learning to Rank in Personal Search" Xuanhui Wang et al.]
+       (https://static.googleusercontent.com/media/research.google.com/ja//pubs/archive/46485.pdf)
+
+       :Parameters:
+
+       alpha : float, default=1
+       Weights associated with regularization term
+
+       max_iter : int, default=100
+       Maximum number of iterations taken for the solvers to converge
+
+       epsilon : float, default=1e-10
+       Tiny value to avoid division by zero.
+
+       class_weights : 'balanced', default = None
+       The "balanced" mode uses the values of y to automatically adjust
+       weights inversely proportional to class frequencies in the input data
+       as ``n_samples / (n_classes * np.bincount(y))``
+       Note that these weights will be multiplied with sample_weight (passed
+       through the fit method) if sample_weight is specified.
+
+       feature_index : int
+       the index to devide X into left latent features and right latent features.
```

Comment:
I think "The first column index of right latent feature." is easier to understand.

Reply:
Fixed it.

Reply:
Fixed it.